File: python/endpoints/localization/convert_xy_to_ll.py
from clearpath_config.common.utils.yaml import read_yaml
from clearpath_localization_msgs.srv import ConvertCartesianToLatLon
from geometry_msgs.msg import PoseStamped, Pose, Point
import rclpy
from rclpy.node import Node

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'ConvertCartesianToLatLon'

# TODO: Enter your Cartesian coordinates
CARTESIAN_MSG = PoseStamped(
    pose=Pose(
        position=Point(
            x=,
            y=,
            z=,
        ),
    )
)


class ConvertCartesianToLatLonNode(Node):

    def __init__(self, namespace: str):
        super().__init__('convert_cartesian_to_lat_lon')
        self._convert_client = self.create_client(ConvertCartesianToLatLon, f'/{namespace}/localization/xy_to_lat_lon')

        while not self._convert_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/localization/xy_to_lat_lon service not available, waiting again...')

    def convert(self, msg):
        req = ConvertCartesianToLatLon.Request()
        req.msg = msg
        self.future = self._convert_client.call_async(req)
        self.future.add_done_callback(self.convert_srv_response_cb)

    def convert_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Converted Cartesian to lat/lon successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to convert Cartesian to lat/lon')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    convert_client = ConvertCartesianToLatLonNode(namespace)
    convert_client.convert(CARTESIAN_MSG)

    rclpy.spin(convert_client)


if __name__ == '__main__':
    main()

File: cpp/endpoints/localization/convert_xy_to_ll.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_localization_msgs/srv/convert_cartesian_to_lat_lon.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

// TODO: Enter your Cartesian coordinates
const double X = 0.0;
const double Y = 0.0;
const double Z = 0.0;

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("convert_xy_to_lat_lon_client");
  rclcpp::Client<clearpath_localization_msgs::srv::ConvertCartesianToLatLon>::SharedPtr client =
    node->create_client<clearpath_localization_msgs::srv::ConvertCartesianToLatLon>("/" + namespace_ + "/" + std::string("localization/xy_to_lat_lon"));

  auto request = std::make_shared<clearpath_localization_msgs::srv::ConvertCartesianToLatLon::Request>();
  auto cartesian_msg = geometry_msgs::msg::PoseStamped();
  cartesian_msg.pose.position.x = X;
  cartesian_msg.pose.position.y = Y;
  cartesian_msg.pose.position.z = Z;
  request->msg = cartesian_msg;

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("convert_xy_to_lat_lon_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("convert_xy_to_lat_lon_client"), "/%s/localization/xy_to_lat_lon service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("convert_xy_to_lat_lon_client"), "Converted xy to lat/lon successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("convert_xy_to_lat_lon_client"), "Failed to convert xy to lat/lon");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("convert_xy_to_lat_lon_client"), "Failed to call service /%s/localization/xy_to_lat_lon", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}
